pymuonsuite/io/castep.py

```diff
@@ -7,109 +7,348 @@
+        # if the params have been changed, the calc has to be remade
+        # from scratch:
+        self._calc = None
+        self._create_calculator()
+
+    def set_script(self, script):
+        '''
+        |   script (str):           Path to a file containing a submission
+        |                           script to copy to the input folder. The
+        |                           script can contain the argument
+        |                           {seedname} in curly braces, and it will
+        |                           be appropriately replaced.
+        '''
+        self.script = script
+
+    def read(self, folder, sname=None):
+        """Reads Castep output files.
+
+        | Args:
+        |   folder (str):           Path to folder from which to read files.
+        |   sname (str):            Seedname to save the files with. If not
+        |                           given, use the name of the folder.
+        """
+        atoms = self._read_castep(folder, sname)
+        self._read_castep_hyperfine_magres(atoms, folder, sname)
+        self._read_castep_gamma_phonons(atoms, folder, sname)
+        return atoms
+
+    def _read_castep(self, folder, sname=None):
+        try:
+            if sname is not None:
+                cfile = os.path.join(folder, sname + '.castep')
+            else:
+                cfile = glob.glob(os.path.join(folder, '*.castep'))[0]
+                sname = seedname(cfile)
+            atoms = io.read(cfile)
+            atoms.info['name'] = sname
+            return atoms
+
+        except IndexError:
+            raise IOError("ERROR: No .castep files found in {}."
+                          .format(os.path.abspath(folder)))
+        except OSError as e:
+            raise IOError("ERROR: {}".format(e))
+        except (io.formats.UnknownFileTypeError, ValueError, TypeError,
+                Exception) as e:
+            raise IOError("ERROR: Invalid file: {file}"
+                          .format(file=sname + '.castep'))
+
+    def _read_castep_hyperfine_magres(self, atoms, folder, sname=None):
+        try:
+            if sname is not None:
+                mfile = os.path.join(folder, sname + '.magres')
+            else:
+                mfile = glob.glob(os.path.join(folder, '*.magres'))[0]
+            m = parse_hyperfine_magres(mfile)
+            atoms.arrays.update(m.arrays)
+        except (IndexError, OSError):
+            warnings.warn("No .magres files found in {}."
+                          .format(os.path.abspath(folder)))
+
+    def _read_castep_gamma_phonons(self, atoms, folder, sname=None):
+        """Parse CASTEP phonon data into a casteppy object,
+        and return eigenvalues and eigenvectors at the gamma point.
+        """
+        try:
+            from euphonic import QpointPhononModes
+        except ImportError:
+            raise ImportError("""
+        Can't use castep phonon interface due to Euphonic not being installed.
```

Comment:
Yes, it is currently in there twice - it appears in `average.py` as well. That makes sense, I'll take a look at how it works in soprano.

Reply:
I made a optional.py file and a function that imports the QpointPhononModes and then when I went to use it in average.py I noticed that it wasn't actually needed there because the code was repeated from castep.py! So in fact we do only need this import once. Seeing as I've already made the optional.py I am inclined to keep it though...

Reply:
The changes are here: https://github.com/muon-spectroscopy-computational-project/pymuon-suite/pull/7/commits/12a6e497697a2d5d2075f8524f9d9c87fa0f2caf . Happy to revert the addition of optional.py if you prefer

Reply:
No, let's keep it. It might come useful in the future anyway.

Reply:
Great!